sgkit/io/bgen/bgen_reader.py

```diff
@@ -0,0 +1,618 @@
+        much slower than subsequent calls.
+    sample_path
+        Path to ``.sample`` file, by default None. This is used to fetch sample identifiers
+        and when provided it is preferred over sample identifiers embedded in the ``.bgen`` file.
+    chunks
+        Chunk size for genotype probability data (3 dimensions),
+        by default "auto".
+    lock
+        Whether or not to synchronize concurrent reads of
+        file blocks, by default False. This is passed through to
+        [dask.array.from_array](https://docs.dask.org/en/latest/array-api.html#dask.array.from_array).
+    persist
+        Whether or not to persist variant information in memory, by default True.
+        This is an important performance consideration as the metadata file for this data will
+        be read multiple times when False.
+    contig_dtype
+        Data type for contig names, by default "str".
+        This may also be an integer type (e.g. "int"), but will fail if any of the contig names
+        cannot be converted to integers.
+    gp_dtype
+        Data type for genotype probabilities, by default "float32".
+
+    Warnings
+    --------
+    Only bi-allelic, diploid BGEN files are currently supported.
+
+    Returns
+    -------
+    A dataset containing the following variables:
+
+    - :data:`sgkit.variables.variant_id` (variants)
+    - :data:`sgkit.variables.variant_contig` (variants)
+    - :data:`sgkit.variables.variant_position` (variants)
+    - :data:`sgkit.variables.variant_allele` (variants)
+    - :data:`sgkit.variables.sample_id` (samples)
+    - :data:`sgkit.variables.call_dosage` (variants, samples)
+    - :data:`sgkit.variables.call_dosage_mask` (variants, samples)
+    - :data:`sgkit.variables.call_genotype_probability` (variants, samples, genotypes)
+    - :data:`sgkit.variables.call_genotype_probability_mask` (variants, samples, genotypes)
+
+    """
+    if isinstance(chunks, tuple) and len(chunks) != 3:
+        raise ValueError(f"`chunks` must be tuple with 3 items, not {chunks}")
+    if not np.issubdtype(gp_dtype, np.floating):
+        raise ValueError(
+            f"`gp_dtype` must be a floating point data type, not {gp_dtype}"
+        )
+    if not np.issubdtype(contig_dtype, np.integer) and np.dtype(
+        contig_dtype
+    ).kind not in {"U", "S"}:
+        raise ValueError(
+            f"`contig_dtype` must be of string or int type, not {contig_dtype}"
+        )
+
+    path = Path(path)
+    sample_path = Path(sample_path) if sample_path else path.with_suffix(".sample")
+
+    if sample_path.exists():
+        sample_id = read_samples(sample_path).sample_id.values.astype("U")
+    else:
+        sample_id = _default_sample_ids(path)
+
+    bgen_reader = BgenReader(path, metafile_path=metafile_path, dtype=gp_dtype)
+
+    df = read_metafile(bgen_reader.metafile_path)
+    if persist:
+        df = df.persist()
+    arrs = dataframe_to_dict(df, METAFILE_DTYPE)
```

Comment:
I would probably use https://github.com/pydata/xarray/issues/3929 if it existed (with default dimension names), but I'm not here to avoid repeating so much of what create_genotype_dosage_dataset does (i.e. the dimension labels don't help in this case).   It's also a reasonably convenient place to put the special string conversion handling.